### PR TITLE
DOC: Add section on debugging Python wrappings

### PR DESCRIPTION
**Pull request by tbirdso**
Add subsection for discussion of ITK Python debugging workflow based on @brad-t-moore 's internal presentation and upcoming blog post.

### CODE REVIEW

**Comment by tbirdso** [comment]
I've verified debugging steps with Python's `pdb` module as well as stepping through ITK files via `RelWithDebInfo` build debug symbols in Visual Studio on Windows. Confirmation of equivalent workflow using `gdb` on Linux is pending, currently waiting for ITK to finish building.

**Comment by tbirdso** [comment]
Pinging @thewtex and @brad-t-moore for review, comments, revisions, etc

**Comment by tbirdso** [comment]
I'm waiting for ITK to build so that I can verify whether the `apt-get install python<version>-dbg` step is actually required on Linux, but otherwise have confirmed that I can step through ITK and SWIG-generated source code on both Windows and Linux using the described process. If this step is not required I will remove it before merge.

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 1]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+
+While the Python debugger is useful, it does not allow us to examine implementation
+details of ITK classes on its own. A deeper investigation may be necessary for localizing
+errors in ITK classes.
+
+
+\subsubsection{SWIG C++ Runtime Tracing}
+
+As discussed in Section~\ref{subsubsec:SwigArchitecture}, SWIG wrapping generates
+C++ source files managing template implementations during ITK compilation.
```
Comment (thewtex) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1012]: managing template implementations -> and manage the interface and ownership semantics between Python objects and C++ objects

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 2]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+When debug symbols are available a running Python process may be attached to step through
```
Comment (thewtex) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1013]: debug symbols -> binary debug symbols
available -> available,

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 3]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+  \item The ITK build must be configured so that debug symbols
+    are generated. Python wrapping must also be enabled.
+    This is accomplished by setting the CMake variables
+    \code{CMAKE\_BUILD\_TYPE:STRING="RelWithDebInfo"} and \code{ITK\_WRAP\_PYTHON:BOOL="On"}.
+    Note that the "RelWithDebInfo" build type is strongly encouraged over a "Debug"
+    build as the former will build against a standard Python distribution.
+    See Section~\ref{sec:UsingCMakeForConfiguringAndBuildingITK} for a detailed
+    explanation of how to build ITK locally with CMake.
+  \item A Python virtual environment must be appropriately configured with
+    \code{WrapITK.pth} so that the ITK debug build can be loaded.
```
Comment (thewtex) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1026]: loaded -> \code{import}'ed

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 4]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+    See Section~\ref{sec:Wrapping} for further explanation.
+  \item ITK modules to be debugged must be loaded in a new Python session initialized
```
Comment (thewtex) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1028]: ITK -> The ITK

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 5]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+>>> import itk
+>>> itk.force_load()
+>>> import os
+>>> os.getpid()
+99999
+\end{minted}
+
+  \item The Python session may be attached to a debugger using the returned PID.
+    \begin{itemize}
+      \item On a Windows operating system Microsoft Visual Studio 2019 or a similar platform
```
Comment (thewtex) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1044]: system -> system,

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 6]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+        from the Debug menu, choose "Native" code, and then search for the process PID. If
+        debug symbols were generated correctly then ITK modules will appear under the list of
+        loaded modules.
+      \item On a Linux operating system the GNU Project Debugger \code{gdb} can be used
+        for attaching to a running Python process, reading symbol files, and setting breakpoints.
+        The following example attaches to a running process and sets a breakpoint inside
+        an \code{itk.Image} Python object.
+
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{bash}
+(venv-itk) > sudo gdb
```
Comment (thewtex) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1055]: No `sudo`, correct?

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 7]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+... continue in Python session until breakpoint is hit ...
+\end{minted}
+
+    \end{itemize}
+
+\end{enumerate}
+
+With these steps completed the respective debugger can be used to set breakpoints
+and step through C++ source code for a respective ITK Python execution.
+
```
Comment (thewtex) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1076]: We could also add a section here on how to get a backtrace when a Python process crashes with a segfault. With Linux / gdb, this is setting `ulimit -c` and invoking `gdb` as described here:

https://pythondev.readthedocs.io/debug_tools.html#create-a-core-dump-file
https://nick-otter.github.io/linux/kernel/memory/2020/11/25/how-to-debug-crashes-and-segfaults/

I am not sure of the process on Windows / macOS, but that could be added if you can find it.

**Review by thewtex — CHANGES_REQUESTED** [review]
@tbirdso fantastic work!

A few suggestions inline.

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 6 above)
Comment (tbirdso) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1055]: The user needs certain permissions to attach to a running process, in my anecdotal experience this always seems to require superuser permissions for attaching to a running Python process.

Trying to attach without `sudo` returns an error, either when attaching with `attach` or attaching at init with `gdb -p <pid>`:
```sh
 (gdb) attach 9216
Attaching to process 9216
Could not attach to process.  If your uid matches the uid of the target
process, check the setting of /proc/sys/kernel/yama/ptrace_scope, or try
again as the root user.  For more details, see /etc/sysctl.d/10-ptrace.conf
ptrace: Operation not permitted.
```

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 7 above)
Comment (tbirdso) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1076]: Great idea, I'll look into the Windows side. I'm limited as to how much I can investigate on macOS, but I imagine the process is very similar to Linux?

**Comment by dzenanz** [comment]
@tbirdso ping me after you amend with Matt's suggestions.

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 7 above)
Comment (thewtex) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1076]: Great, thanks! @dzenanz or @brad-t-moore may know the incantations for Windows.

macOS is similar, although the standard debugger is now `lldb`. You can ssh into `misty` if you need a system to experiment on.

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 6 above)
Comment (thewtex) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1055]: Interesting, was the `python` process started with the same `uid`?

Comment (tbirdso) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1055]: Yes, I checked back to confirm that both the `python` and `gdb` processes contain a `uid`. Looking around I can't find any other documentation stating superuser permissions as a requirement, though. I'll amend to remove `sudo` and mention that elevated permissions *may* be required.

**Comment by jhlegarreta** [comment]
Have not gone through the details/would not be able to test them, and sorry to chime in. Thanks for doing this @tbirdso :1st_place_medal: . A few of notes:
- I am not sure about how consistent we have been indenting the `LaTeX` `enumerate`/`itemize`, lists, but I think it is preferable not to indent them for the sake of readability.
- I'd advocate for adding labels to all `section`/`subsection`/`subsubsection`, etc. categories. Even if at the moment we may not need to cross-ref them, they might be useful at some point.
- I'd (try to) provide a more _development_ or _architecture_ vs. _debugging_ focus to the content, maybe slightly rephrasing section names or re-thinking some of the wording used. As it stands now (e.g. _Strategies for Resolving Issues in Wrapped Classes_, and mentions to _debugger_ -I confess that this last might be necessary), parts of it might be more fit for a section in ITK's `CONTRIBUTING.md`, as some sort of FAQ, debugging (i.e. resolving issues), or troubleshooting section**. So maybe slightly re-wording the relevant section names or used wording might help in providing the necessary development (vs. debugging) focus to the section. I acknowledge that it might be hard to do this.

** I confess that other sections might exist in the ITK SW Guide that are also intended for similar purposes (most notably, the git workflow one, that's my fault).

**Comment by tbirdso** [comment]
Thanks for weighing in @jhlegarreta, glad to have your feedback!
- I can remove indentations and add labels to the sections in the scope of this PR. There is some inconsistency in how this is handled throughout the file, but I think that's best addressed in another PR if at all, given that it doesn't translate to the final PDF.
- You're right, it is hard to balance the architecture vs debugging discussion here as neither one is quite as useful without the other. This section gets deeper into the weeds than `CONTRIBUTING.md` currently does, and I would argue that the level of detail required to explain wrapping debugging steps makes the software guide a better fit, at least in the absence of a dedicated troubleshooting chapter. I'd be happy to make a similar debugging post on the community discourse where active ITK troubleshooting is happening, though codifying this section in the software guide will make sure it doesn't get lost or buried over time. There's also a technical blog post in the works to help disseminate this information for developers. I'll think on how we can change section titles to fit better with the rest of the guide.

**Comment by jhlegarreta** [comment]
> I can remove indentations and add labels to the sections in the scope of this PR. There is some inconsistency in how this is handled throughout the file, but I think that's best addressed in another PR if at all, given that it doesn't translate to the final PDF.

OK.

>  I'd be happy to make a similar debugging post on the community discourse where active ITK troubleshooting is happening

Not necessary IMHO. A link to the section or eventual blog post would suffice in discourse to avoid redundancy.

> though codifying this section in the software guide will make sure it doesn't get lost or buried over time.

👌.

 > There's also a technical blog post in the works to help disseminate this information for developers.

That's awesome 💯.

**Comment by thewtex** [comment]
@hjmjohnson and @seanm are macOS experts and may have guidance for attaching or debugging core dumps.

**Review by thewtex — DISMISSED** [review]

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 8]
```diff
@@ -948,27 +948,34 @@ \subsubsection{Swig Python Architecture}
...
+Debug symbols will be generated for these files.
+
+\item SWIG compiled code. For Python wrappings these are generated as Python
+\code{.pyd} (Windows) or \code{.so} (Linux or macOS) binaries and Python \code{.py} modules at
+\code{Wrapping/Generators/Python/itk} in the ITK build tree.
+
+\item Additional Python configuration files are generated in the
+\code{Wrapping/Generators/Python} directory and its subdirectories.
+\code{WrapITK.pth} provides the path for a Python environment to find the ITK
+module, while \code{\_\_init\_\_.py} allows the module during development to be loaded correctly at
```
Comment (dzenanz) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:968]: during -> under

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 9]
```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
...
+\end{minted}
+
+    \end{itemize}
+
+\end{enumerate}
+
+With these steps completed the respective debugger can be used to set breakpoints
+and step through C++ source code for a respective ITK Python execution.
+
+
```
Comment (dzenanz) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1093]:
```suggestion
In Visual Studio, debugging can be enabled right from the start if the Python script is loaded into Visual Studio as part of a "Python Project". Mixed mode debugging needs to be enabled, as per documentation:
\url{https://docs.microsoft.com/en-us/visualstudio/python/debugging-mixed-mode-c-cpp-python-in-visual-studio?view=vs-2019}
Starting a project like this is an order of magnitude slower, as many debug symbols need to be loaded and examined.
```

Comment (dzenanz) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1093]: I don't know whether `\url{...}` is the right syntax.

**Comment by brad-t-moore** [comment]
Sorry for my ignorance here- is there a way to see the generated output through git or do I need to checkout this PR and build locally?

-Brad

**Comment by tbirdso** [comment]
Good question @brad-t-moore , you can download the PDFs generated by CI under "Artifacts" from the run:
[https://github.com/InsightSoftwareConsortium/ITKSoftwareGuide/actions/runs/1554481810](https://github.com/InsightSoftwareConsortium/ITKSoftwareGuide/actions/runs/1554481810)

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 7 above)
Comment (tbirdso) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1092]: It looks like Visual Studio makes the program call stack, locals, etc available on breakpoints or crashes, and the user can manually create and then reload dumps as desired. Process is described here, I will add a summary and URL reference: [https://docs.microsoft.com/en-us/visualstudio/debugger/using-dump-files?view=vs-2022](https://docs.microsoft.com/en-us/visualstudio/debugger/using-dump-files?view=vs-2022)

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 9 above)
Comment (tbirdso) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex]: I believe `\url` is the right syntax, thanks for contributing!

**Comment by tbirdso** [comment]
For the record, I didn't knowingly "dismiss" anything...

**Review by dzenanz — APPROVED** [review]

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 7 above)
Comment (tbirdso) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1094]: As a status update here, I'm having trouble attaching LLDB to a Python process on `misty`:
```sh
error: process exited with status -1 (this is a non-interactive debug session, cannot get permission to debug processes.)
```

This occurs for an interactive LLDB session for both interactive and non-interactive Python sessions. LLDB seems to attach to other processes just fine. Running with elevated permissions does not resolve the issue, and I haven't been able to find much discussion online. The machine is in developer mode. Would welcome discussion from macOS users who may have seen this before  😃

**Comment by tbirdso** [comment]
Added MacOS debugging steps (thanks to @thewtex for the use of the misty machine and @timthirion for the brief use of his laptop 😃 )

@hjmjohnson @seanm If you have an opportunity I would welcome your review on the MacOS side, am happy to revise based on your development experience.

@brad-t-moore please also let me know if you have thoughts/revisions on the debugging process per your own investigations

Other reviews welcome as always. Thanks!

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 10]
```diff
@@ -1085,19 +1091,80 @@ \subsubsection{C++ Runtime Tracing}
...
+DevToolsSecurity -enable
+\end{minted}
+
+\item It may be necessary to add entitlements to the Python executable so that
+\code{lldb} can attach to the hardened process. This may be accomplished by
+updating a \code{.plist} entitlements file and setting the executable entitlements.
+
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{bash}
+codesign -d --entitlements :- "/path/to/python" >> "/tmp/path/to/python_entitlements.plist"
+/usr/libexec/PlistBuddy -c "Add :com.apple.security.get-task-allow bool true" "/tmp/path/to/python_entitlements.plist"
```
Comment (dzenanz) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1150]: Are long lines wrapped in the generated PDF, or do they go off the page?

Comment (tbirdso) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:1150]: Good thought, looking back at the [artifacts from the previous run](https://github.com/InsightSoftwareConsortium/ITKSoftwareGuide/actions/runs/1579072802) there are several instances of lines running off the page. Will fix.

**Review by dzenanz — DISMISSED** [review]

**Comment by tbirdso** [comment]
Squashed changes.

**Review by dzenanz — APPROVED** [review]

**Comment by tbirdso** [comment]
I do not have write access to this repo -- could @dzenanz or @thewtex merge when ready?

**Comment by tbirdso** [comment]
Thank you @dzenanz !

**Comment by thewtex** [comment]
:bug: :champagne: :santa:

**Pull request merged**